### PR TITLE
SOLR-16487: Improve Pull Replica Interval

### PR DESCRIPTION
**Pull request by justinrsweeney**
https://issues.apache.org/jira/browse/SOLR-16487

# Description

The polling for replication is fairly simply, but can lead to polling too often. As an example if you had the following config for commits:
```
<autoCommit>
    <maxTime>15000</maxTime>
    <openSearcher>false</openSearcher>
</autoCommit>

<autoSoftCommit>
    <maxTime>60000</maxTime>
</autoSoftCommit>
```
The current logic would setup polling to be half of the autoCommit time, so poll every 7.5 seconds. However since a new searcher isn't opened, there will only be changes reflected every 60 seconds on the leader. We can make this logic a bit smarter knowing that the replication handler won't reflect changes until a new searcher is opened.

# Solution

Modified the code setting the polling interval for replication to only use the autoCommit time if openSearcher is true, otherwise it will use the soft commit time.

# Tests

Please describe the tests you've developed or run to confirm this patch implements the feature or solves the problem.

# Checklist

Please review the following and check all that apply:

- [x] I have reviewed the guidelines for [How to Contribute](https://wiki.apache.org/solr/HowToContribute) and my code conforms to the standards described there to the best of my ability.
- [x] I have created a Jira issue and added the issue ID to my pull request title.
- [ ] I have given Solr maintainers [access](https://help.github.com/en/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork) to contribute to my PR branch. (optional but recommended)
- [x] I have developed this patch against the `main` branch.
- [x] I have run `./gradlew check`.
- [x] I have added tests for my changes.
- [ ] I have added documentation for the [Reference Guide](https://github.com/apache/solr/tree/main/solr/solr-ref-guide)

### CODE REVIEW

File: solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java [hunk 1]
```diff
@@ -155,6 +155,38 @@ public static String getCommitVersion(SolrCore solrCore) {
...
+        // softCommit is configured.
+        // Usually if softCommit is configured, `hardCommitNewSearcher==false`, in which case you
+        // want
+        // to calculate poll interval wrt the max of hardCommitTime (when segments are available to
+        // replicate) and softCommitTime (when changes are visible).
+        // But in the unusual case that hardCommit _does_ open a new searcher and
+        // `hardCommitMaxTime < softCommitMaxTime`, then fallback to `else` clause, setting poll
+        // interval
+        // wrt `hardCommitMaxTime` alone.
+        pollIntervalStr = toPollIntervalStr((Math.max(hardCommitMaxTime, softCommitMaxTime)) / 2);
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java:178]: nitpick: extra parentheses in here that are not adding clarity at all

File: solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java [hunk 2]
```diff
@@ -155,6 +155,38 @@ public static String getCommitVersion(SolrCore solrCore) {
...
   }
 
+  public static String determinePollInterval(SolrConfig.UpdateHandlerInfo uinfo) {
+    int hardCommitMaxTime = uinfo.autoCommmitMaxTime;
+    int softCommitMaxTime = uinfo.autoSoftCommmitMaxTime;
+    boolean hardCommitNewSearcher = uinfo.openSearcher;
+    String pollIntervalStr = null;
+    if (hardCommitMaxTime != -1) {
+      // configured hardCommit places a ceiling on the interval at which new segments will be
+      // available
```
Comment (dsmiley) [inline solr/core/src/java/org/apache/solr/cloud/ReplicateFromLeader.java:165]: Code comments need to be re-flowed after "tidy" did its work.
When writing comments that flow to this many lines (kind of a lot) consider /* */ style, which I *think* Tidy reflows for you.

File: solr/core/src/test/org/apache/solr/cloud/ReplicateFromLeaderTest.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.solr.cloud;
+
+import org.apache.solr.core.SolrConfig;
+import org.junit.Test;
+
+public class ReplicateFromLeaderTest extends SolrCloudTestCase {
```
Comment (dsmiley) [inline solr/core/src/test/org/apache/solr/cloud/ReplicateFromLeaderTest.java:23]: Why SolrCloudTestCase?

Comment (justinrsweeney) [reply solr/core/src/test/org/apache/solr/cloud/ReplicateFromLeaderTest.java]: No good reason, thanks for pointing that out.

**Pull request merged**